### PR TITLE
Added a signal when the link status changes

### PR DESCRIPTION
**Pull request by rohithasrk**
This PR references #27. As of now, only a test signal has been added to confirm if the implementation of django-signals and the process of sending signals is right.

### CODE REVIEW

File: django_netjsongraph/base/link.py [hunk 1]
```diff
@@ -73,6 +86,9 @@ def json(self, dict=False, **kwargs):
             return netjson
         return json.dumps(netjson, cls=JSONEncoder, **kwargs)
 
+    def send_signal(self):
+        self.link_status_change.send(sender=self.__class__, topology=self.topology, id=self.pk)
```
Comment (nemesisdesign) [inline django_netjsongraph/base/link.py]: send the Link instance

File: django_netjsongraph/base/link.py [hunk 2]
```diff
@@ -37,9 +38,21 @@ class AbstractLink(TimeStampedEditableModel):
...
                            dump_kwargs={'indent': 4})
 
+    link_status_change = Signal(providing_args=["topology", "id"])
+
     class Meta:
         abstract = True
 
+    def __init__(self, *args, **kwargs):
+        super(AbstractLink, self).__init__(*args, **kwargs)
+        self.initial_status = self.status
```
Comment (nemesisdesign) [inline django_netjsongraph/base/link.py]: flag it as private, eg: `self._initial_status`

File: django_netjsongraph/base/link.py [hunk 3]
```diff
@@ -37,9 +38,21 @@ class AbstractLink(TimeStampedEditableModel):
...
+
+    def save(self, *args, **kwargs):
+        if self.status != self.initial_status:
```
Comment (nemesisdesign) [inline django_netjsongraph/base/link.py]: move it after the `save()` call, so the signal is sent after the status has changed

File: django_netjsongraph/base/link.py [hunk 4]
```diff
@@ -37,9 +38,21 @@ class AbstractLink(TimeStampedEditableModel):
                            load_kwargs={'object_pairs_hook': OrderedDict},
                            dump_kwargs={'indent': 4})
 
+    link_status_change = Signal(providing_args=["topology", "id"])
```
Comment (nemesisdesign) [inline django_netjsongraph/base/link.py]: I think signals are not usually defined as class attributes, so I think you can move this before the model definition.

Also, I would prefer `link_status_changed`.

File: django_netjsongraph/base/link.py [hunk 5]
```diff
@@ -73,6 +86,9 @@ def json(self, dict=False, **kwargs):
             return netjson
         return json.dumps(netjson, cls=JSONEncoder, **kwargs)
 
+    def send_signal(self):
```
Comment (nemesisdesign) [inline django_netjsongraph/base/link.py]: rename to `send_status_changed_signal`

File: django_netjsongraph/base/link.py (hunk 3 above)
Comment (rohithasrk) [inline django_netjsongraph/base/link.py]: Okay.

File: django_netjsongraph/base/link.py (hunk 1 above)
Comment (rohithasrk) [inline django_netjsongraph/base/link.py]: Okay. Sure!

File: django_netjsongraph/tests/test_link.py [hunk 6]
```diff
@@ -83,3 +84,21 @@ def test_get_from_nodes(self):
...
+        self.signal_was_called = False
+        l = Link(topology_id="a083b494-8e16-4054-9537-fb9eba914861",
+                 source_id="d083b494-8e16-4054-9537-fb9eba914861",
+                 target_id="d083b494-8e16-4054-9537-fb9eba914862",
+                 cost=1.0,
+                 status='up')
+        l.save()
+
+        def handler(sender, link, **kwargs):
+            self.signal_was_called = True
```
Comment (nemesisdesign) [inline django_netjsongraph/tests/test_link.py]: Add a few paranoid assertions:

self.assertEqual(link.pk, l.pk)
self.assertEqual(link.status, 'down')

**Pull request closed without merging**